### ISSUE TITLE
usePureState

### ISSUE THREAD

**Issue by cloudever** [streamich/react-use#60]
Use the same as useState but with shallow compare

**Comment by streamich** [comment]
Do you mean it would not re-render if props didn't change?

**Comment by revskill10** [comment]
Could `react-fast-compare` help here ?

**Comment by streamich** [comment]
`react-fast-compare` does deep compare, as I understand, not shallow.

**Comment by cloudever** [comment]
I mean that if use useState with object like this.setState, it would be great to skip update using shallow compare of object values like that

```js
function Component() {
  const [state, setState] = usePureState({ a: 1, b: 2 })

  // skip update
  setState({ a: 1 }) // merges automatically
  // updates
  setState({ a: 2})
 
}
```

**Comment by streamich** [comment]
Yeah, sounds like something useful. I guess, could be easy to do, too, just with just a single loop that shallow-checks all values.

**Comment by chriswilty** [comment]
This is how React do shallowEquals:
https://github.com/facebook/react/blob/master/packages/shared/shallowEqual.js

I would hope that usePureState could get included as a built-in hook. Saves us all re-implementing (read _copy-pasting_) the above comparator.

**Comment by dpikt** [comment]
Could this be implemented via a hook that allows a custom comparator function to be used for state updates? It looks like `useState` already avoids re-renders for shallowly equal updates, we would just need to expand the concept of what an "equal" update is.

```js
// using default useState comparator 
const [ state, setState ] = useDistinctState('a')
setState('A') // will update
...
// using a custom comparator
const [ state, setState ] = useDistinctState('a', 
    (prev, next) => prev.toUpperCase() === next.toUpperCase()
)
setState('A') // won't update
...
// solving issue use case
function usePureState (initialValue) {
   return useDistinctState(initialValue, shallowEqual)
}
```
(I just came from reading #92 so that's part of my consideration here.)
